File: GameEngine_IOS/GameEngine/ModelGenerators/WorldGUIsGenerator.swift
import Foundation

/**
 * Responsible for creating the multiple GUIs to the user interact with 3D world
 */
public class WorldGUIsGenerator {
    
    /*Zoom that is going apply in the buttons of the gam*/
    private static let BUTTONS_ZOOM : Float = 0.07;
    
    /*Position of the buttons that are upper in the Y*/
    private static let UPPER_BUTTONS : Float = -0.7;
    
    /*Position of the buttons that are in the middle in the Y*/
    private static let MIDDLE_BUTTONS : Float = -0.8;
    
    /*Position of the buttons that are upper in the Y*/
    private static let BOTTOM_BUTTONS : Float = -0.9;
    
    /**
     *
     * @param loader
     *            The loader to load the texture of the GUI
     * @param rawMode
     *            Shape of the GUI
     * @param textureFileName
     *            The name of the file where the texture exists
     * @param xPosition
     *            xPosition of the GUI
     * @param yPosition
     *            yPosition of the GUI
     * @param scale
     *            The scale of the GUI
     * @param key               The key that GUI match if any
     *
     * @return The textured GUI to render GUI
     */
    private static func  getGUI(loader : Loader, _ rawMode : RawModel, _ textureFileName : String, _ xPosition : Float,
                                _ yPosition : Float, _ scale : Float, _ key :GamePadKey!) -> GuiTexture {
        // Load the texture of the GUI
        let textureId : Int = loader.loadTexture(textureFileName);
        // Create the position of the GUI
        let guiPosition : Vector2f = Vector2f(x: xPosition, y: yPosition);
        // The scale vector
        let guiScale : Vector2f  = Vector2f(x: scale, y: scale);
        // Create the textured GUI
        let guiTexture : GuiTexture = GuiTexture(rawMode, textureId, guiPosition, guiScale, key);
        
        return guiTexture;
    }
    
    /**
     * The GUIs of the scene
     *
     * @param loader
     *            The loader in charge of loading the textures of the terrains
     *
     * @return list of terrains of the scene
     */
    public static func getGUIs(loader : Loader) -> Array<GuiTexture> {
        let guiShape : GuiShape = GuiShape();
        let rawModel : RawModel = loader.load2DPositionsToVAO(guiShape.getVertices(),positionsLength: Int(guiShape.countVertices()));
        
        let leftButton = getGUI(loader, rawModel, "ic_pad_left", -0.8, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.left);
        let upButton = getGUI(loader, rawModel, "ic_pad_up", -0.7, UPPER_BUTTONS, BUTTONS_ZOOM, GamePadKey.up);
        let downButton = getGUI(loader, rawModel, "ic_pad_down", -0.7, BOTTOM_BUTTONS, BUTTONS_ZOOM, GamePadKey.down);
        let rightButton = getGUI(loader, rawModel, "ic_pad_right", -0.6, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.right);
        let squareButton = getGUI(loader, rawModel, "ic_pad_left", 0.6, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.square);
        let xButton = getGUI(loader, rawModel, "ic_pad_down", 0.7, BOTTOM_BUTTONS, BUTTONS_ZOOM, GamePadKey.x);
        let triangleButton = getGUI(loader, rawModel, "ic_pad_up", 0.7, UPPER_BUTTONS, BUTTONS_ZOOM, GamePadKey.triangle);
        let circleButton = getGUI(loader, rawModel, "ic_pad_right", 0.8, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.circle);

        
        let guis : Array<GuiTexture> = [leftButton,upButton,downButton, rightButton, squareButton, xButton, triangleButton, circleButton];
        return guis;
    }
}

File: GameEngine_Android/app/src/main/java/com/dferreira/game_engine/modelGenerators/WorldGUIsGenerator.java
package com.dferreira.game_engine.modelGenerators;

import com.dferreira.commons.Vector2f;
import com.dferreira.commons.generic_render.ILoaderRenderAPI;
import com.dferreira.commons.generic_render.IRawModel;
import com.dferreira.game_controller.GamePadKey;
import com.dferreira.game_engine.R;
import com.dferreira.game_engine.models.GuiShape;
import com.dferreira.game_engine.models.GuiTexture;

/**
 * Responsible for creating the multiple GUIs to the user interact with 3D world
 */
public class WorldGUIsGenerator {
    private final static int NUMBER_OF_GUI = 8;

    /*Zoom that is going apply in the buttons of the gam*/
    private final static float BUTTONS_ZOOM = 0.07f;

    /*Position of the buttons that are upper in the Y*/
    private final static float UPPER_BUTTONS = -0.7f;

    /*Position of the buttons that are in the middle in the Y*/
    private final static float MIDDLE_BUTTONS = -0.8f;

    /*Position of the buttons that are upper in the Y*/
    private final static float BOTTOM_BUTTONS = -0.9f;


    /**
     * @param rawMode           Shape of the GUI
     * @param textureResourceId The name of the file where the texture exists
     * @param xPosition         xPosition of the GUI
     * @param yPosition         yPosition of the GUI
     * @param scale             The scale of the GUI
     * @param key               The key that GUI match if any
     * @return The textured GUI to render GUI
     */
    @SuppressWarnings("SameParameterValue")
    private static GuiTexture getGUI(IRawModel rawMode, int textureResourceId, float xPosition,
                                     float yPosition, float scale, GamePadKey key) {
        // Create the position of the GUI
        Vector2f guiPosition = new Vector2f(xPosition, yPosition);
        // The scale vector
        Vector2f guiScale = new Vector2f(scale, scale);
        // Create the textured GUI

        return new GuiTexture(rawMode, textureResourceId, guiPosition, guiScale, key);
    }

    /**
     * The GUIs of the scene
     *
     * @param loaderRenderAPI The loader in charge of loading the textures of the guis
     * @return list of guis of the scene
     */
    public static GuiTexture[] getGUIs(ILoaderRenderAPI loaderRenderAPI) {
        GuiShape guiShape = new GuiShape();
        IRawModel rawModel = loaderRenderAPI.load2DPositionsToRawModel(guiShape.getVertices());
        GuiTexture leftButton = getGUI(rawModel, R.drawable.ic_pad_left, -0.8f, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.left);
        GuiTexture upButton = getGUI(rawModel, R.drawable.ic_pad_up, -0.7f, UPPER_BUTTONS, BUTTONS_ZOOM, GamePadKey.up);
        GuiTexture downButton = getGUI(rawModel, R.drawable.ic_pad_down, -0.7f, BOTTOM_BUTTONS, BUTTONS_ZOOM, GamePadKey.down);
        GuiTexture rightButton = getGUI(rawModel, R.drawable.ic_pad_right, -0.6f, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.right);
        GuiTexture squareButton = getGUI(rawModel, R.drawable.ic_pad_left, 0.6f, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.square);
        GuiTexture xButton = getGUI(rawModel, R.drawable.ic_pad_down, 0.7f, BOTTOM_BUTTONS, BUTTONS_ZOOM, GamePadKey.x);
        GuiTexture triangleButton = getGUI(rawModel, R.drawable.ic_pad_up, 0.7f, UPPER_BUTTONS, BUTTONS_ZOOM, GamePadKey.triangle);
        GuiTexture circleButton = getGUI(rawModel, R.drawable.ic_pad_right, 0.8f, MIDDLE_BUTTONS, BUTTONS_ZOOM, GamePadKey.circle);

        GuiTexture[] GUIs = new GuiTexture[NUMBER_OF_GUI];

        int GUIIndex = 0;
        GUIs[GUIIndex++] = leftButton;
        GUIs[GUIIndex++] = upButton;
        GUIs[GUIIndex++] = downButton;
        GUIs[GUIIndex++] = rightButton;
        GUIs[GUIIndex++] = squareButton;
        GUIs[GUIIndex++] = xButton;
        GUIs[GUIIndex++] = triangleButton;
        GUIs[GUIIndex] = circleButton;

        return GUIs;
    }

    /**
     * Loads the textures of the guis
     *
     * @param loaderRenderAPI Loader to load the raw model
     * @param GUIs            List of the GUIs in the scene
     */
    public static void loadTextures(ILoaderRenderAPI loaderRenderAPI, GuiTexture[] GUIs) {
        if ((GUIs != null) && (GUIs.length != 0)) {
            for (GuiTexture guiTexture : GUIs) {
                // Load the texture of the GUI
                Integer textureId = loaderRenderAPI.loadTexture(guiTexture.getTextureResourceId(), false);
                guiTexture.setTextureId(textureId);
            }
        }
    }
}